File: backend/app/services/stage_indicative.py
"""Derive task/component dates from sub-stage planned and completed dates."""

from __future__ import annotations

from datetime import date

from sqlalchemy.orm import Session

from app.models import ComponentSubStage, ProjectComponent, Task, TaskSubStage


def _stage_start(stage: TaskSubStage | ComponentSubStage) -> date | None:
    return stage.start_date or stage.end_date or stage.due_date


def _stage_end(stage: TaskSubStage | ComponentSubStage) -> date | None:
    return stage.end_date or stage.due_date or stage.start_date


def indicative_dates_from_stages(
    stages: list[TaskSubStage] | list[ComponentSubStage],
) -> tuple[date | None, date | None]:
    starts = [start for s in stages if (start := _stage_start(s))]
    ends = [end for s in stages if (end := _stage_end(s))]
    return (
        min(starts) if starts else None,
        max(ends) if ends else None,
    )


def indicative_dates_from_stage_outs(stages) -> tuple[date | None, date | None]:
    """Compute indicative range from API sub-stage objects (SubStageOut)."""
    starts = [s.start_date or s.end_date or s.due_date for s in stages]
    starts = [d for d in starts if d]
    ends = [s.end_date or s.due_date or s.start_date for s in stages]
    ends = [d for d in ends if d]
    return (
        min(starts) if starts else None,
        max(ends) if ends else None,
    )


def actual_dates_from_completed_stages(
    stages: list[TaskSubStage] | list[ComponentSubStage],
) -> tuple[date | None, date | None]:
    done = [s for s in stages if s.is_done]
    if not done:
        return None, None
    starts = [start for s in done if (start := _stage_start(s))]
    ends = [end for s in done if (end := _stage_end(s))]
    return (
        min(starts) if starts else None,
        max(ends) if ends else None,
    )


def _apply_actual_dates(target: Task | ProjectComponent, start: date | None, end: date | None) -> None:
    target.start_date = start
    target.end_date = end
    if start and end:
        target.duration_days = (end - start).days + 1
    else:
        target.duration_days = None


def recompute_component_indicative_dates(db: Session, component: ProjectComponent) -> None:
    stages = (
        db.query(ComponentSubStage)
        .filter(ComponentSubStage.component_id == component.id)
        .all()
    )
    ind_start, ind_end = indicative_dates_from_stages(stages)
    component.indicative_start = ind_start
    component.indicative_end = ind_end


def recompute_component_actual_dates(db: Session, component: ProjectComponent) -> None:
    stages = (
        db.query(ComponentSubStage)
        .filter(ComponentSubStage.component_id == component.id)
        .all()
    )
    start, end = actual_dates_from_completed_stages(stages)
    _apply_actual_dates(component, start, end)


def recompute_indicative_dates(db: Session, task: Task) -> None:
    if task.component_id and task.component:
        recompute_component_indicative_dates(db, task.component)
        return
    stages = db.query(TaskSubStage).filter(TaskSubStage.task_id == task.id).all()
    ind_start, ind_end = indicative_dates_from_stages(stages)
    task.indicative_start = ind_start
    task.indicative_end = ind_end


def recompute_actual_dates(db: Session, task: Task) -> None:
    if task.component_id and task.component:
        recompute_component_actual_dates(db, task.component)
        return
    stages = db.query(TaskSubStage).filter(TaskSubStage.task_id == task.id).all()
    start, end = actual_dates_from_completed_stages(stages)
    _apply_actual_dates(task, start, end)

File: frontend/src/utils/stageIndicative.ts
import type { SubStage } from '../types'
import { parseDate } from './dateShift'
import { stageEffectiveEndDate } from './subStageDates'

function stageEffectiveStartDate(stage: SubStage): string | null {
  return stage.start_date ?? stage.end_date ?? stage.due_date
}

function inclusiveDaysBetween(start: string, end: string): number {
  const a = parseDate(start)
  const b = parseDate(end)
  if (!a || !b) return 0
  return Math.round((b.getTime() - a.getTime()) / 86400000) + 1
}

/** Min start and max end across completed sub-stages (is_done). */
export function actualDatesFromCompletedStages(stages: SubStage[]): {
  start: string | null
  end: string | null
} {
  const done = stages.filter((s) => s.is_done)
  if (!done.length) return { start: null, end: null }

  const starts: string[] = []
  const ends: string[] = []
  for (const stage of done) {
    const start = stageEffectiveStartDate(stage)
    const end = stageEffectiveEndDate(stage)
    if (start) starts.push(start)
    if (end) ends.push(end)
  }

  return {
    start: starts.length ? starts.reduce((min, d) => (d < min ? d : min)) : null,
    end: ends.length ? ends.reduce((max, d) => (d > max ? d : max)) : null,
  }
}

export function completionPctFromStages(stages: SubStage[]): number {
  if (!stages.length) return 0
  const done = stages.filter((s) => s.is_done).length
  return Math.round((done / stages.length) * 100)
}

export function durationDaysFromStageRange(
  start: string | null,
  end: string | null
): number | null {
  if (!start || !end) return null
  return inclusiveDaysBetween(start, end)
}

/** Min start and max end across planned sub-stages (is_indicative). */
export function indicativeDatesFromStages(stages: SubStage[]): {
  start: string | null
  end: string | null
} {
  const planned = stages.filter((s) => s.is_indicative)
  if (!planned.length) return { start: null, end: null }

  const starts: string[] = []
  const ends: string[] = []
  for (const stage of planned) {
    const start = stageEffectiveStartDate(stage)
    const end = stageEffectiveEndDate(stage)
    if (start) starts.push(start)
    if (end) ends.push(end)
  }

  return {
    start: starts.length ? starts.reduce((min, d) => (d < min ? d : min)) : null,
    end: ends.length ? ends.reduce((max, d) => (d > max ? d : max)) : null,
  }
}
